### PR TITLE
Fixed #26951 -- Fixed AuthenticationForm to work with integer usernames.

### PR DESCRIPTION
**Pull request by alexyer**
https://code.djangoproject.com/ticket/26951

### CODE REVIEW

**Comment by charettes** [comment]
Buildbot, add to whitelist.

File: tests/auth_tests/test_forms.py [hunk 1]
```diff
@@ -351,6 +352,23 @@ def test_password_whitespace_not_stripped(self):
         form.is_valid()  # Not necessary to have valid credentails for the test.
         self.assertEqual(form.cleaned_data['password'], data['password'])
 
+    @override_settings(AUTH_USER_MODEL='auth_tests.CustomUserWithIntegerUsername')
+    def test_integer_field(self):
+        class CustomAuthenticationForm(AuthenticationForm):
+            class Meta(UserCreationForm.Meta):
```
Comment (berkerpeksag) [inline tests/auth_tests/test_forms.py]: `UserCreationForm.Meta` looks like a typo to me. Perhaps you meant `AuthenticationForm.Meta`?

**Comment by alexyer** [comment]
Updated tests. Thanks, @berkerpeksag!

File: tests/auth_tests/models/with_integer_username.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+class CustomUserWithIntegerUsername(AbstractBaseUser):
+    objects = CustomUserWithIntegerUsernameManager()
+
+    username = models.IntegerField()
+    password = models.CharField(max_length=255)
+
+    USERNAME_FIELD = 'username'
+    REQUIRED_FIELDS = ['username', 'password']
+
+    def __str__(self):
```
Comment (berkerpeksag) [inline tests/auth_tests/models/with_integer_username.py]: I think `__str__` can be omitted here. If it needs to stay, `CustomUserWithIntegerUsername` should be wrapped by `@python_2_unicode_compatible`.

File: tests/auth_tests/models/with_integer_username.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        user = self.model(username=username)
+        user.set_password(password)
+        user.save(using=self._db)
+        return user
+
+    def get_by_natural_key(self, username):
+        return self.get(username=username)
+
+
+class CustomUserWithIntegerUsername(AbstractBaseUser):
```
Comment (berkerpeksag) [inline tests/auth_tests/models/with_integer_username.py]: Nitpick: You may want to add `CustomUserWithIntegerUsername` to `__all__` in `tests/auth_tests/models/__init__.py`.

File: django/contrib/auth/forms.py [hunk 4]
```diff
@@ -186,7 +186,7 @@ def clean(self):
         username = self.cleaned_data.get('username')
         password = self.cleaned_data.get('password')
 
-        if username and password:
+        if username is not None and password is not None:
```
Comment (berkerpeksag) [inline django/contrib/auth/forms.py]: This is probably not that important, but the `password is not None` part is not tested.

File: tests/auth_tests/test_forms.py [hunk 5]
```diff
@@ -351,6 +352,22 @@ def test_password_whitespace_not_stripped(self):
...
 
+    @override_settings(AUTH_USER_MODEL='auth_tests.CustomUserWithIntegerUsername')
+    def test_integer_field(self):
+        class CustomAuthenticationForm(AuthenticationForm):
+            username = IntegerField()
+
+        CustomUserWithIntegerUsername.objects.create_user(username=0, password='pwd')
+        data = {
+            'username': 0,
+            'password': 'pwd'
```
Comment (berkerpeksag) [inline tests/auth_tests/test_forms.py]: Nitpick: Could you add a comma after `'password': 'pwd'`?

File: tests/auth_tests/models/with_integer_username.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    objects = CustomUserWithIntegerUsernameManager()
```
Comment (berkerpeksag) [inline tests/auth_tests/models/with_integer_username.py]: Another nitpick: Could you move `objects` to below `username` and `password`?

**Comment by alexyer** [comment]
Updated code style. Thanks, @berkerpeksag!

**Comment by timgraham** [comment]
The test isn't working as a regression test because it passes before the fix is applied. I added one line to fix it: `self.assertEqual(form.user_cache, user)`.

Merged in 975a76a96428bcc9da43d87e3f8433f498da2d68, thanks!

**Pull request closed without merging**